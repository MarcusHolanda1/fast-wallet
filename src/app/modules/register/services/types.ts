import { Card } from '../../../shared/types/card';

export interface CardsResponse {
  cards: Card[];
}
