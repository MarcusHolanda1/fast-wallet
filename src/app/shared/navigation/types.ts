import { Card } from '../types/card';

export type RootStackParamList = {
  HomeScreen: undefined;
  RegisterScreen: undefined;
  RegisterSuccessScreen: {
    card: Card;
  };
  WalletScreen: undefined;
};
