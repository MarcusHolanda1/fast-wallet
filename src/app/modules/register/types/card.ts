export interface Card {
  id: string;
  number: string;
  cvv: string;
  name: string;
}

export interface CardPayload {
  number: string;
  cvv: string;
  name: string;
}
