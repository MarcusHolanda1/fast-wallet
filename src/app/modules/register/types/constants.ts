export interface CardValidatorMessages {
  cardNumber: {
    required: string;
    matches: string;
  };
  cardHolder: {
    required: string;
    min: string;
  };
  expiryDate: {
    required: string;
    matches: string;
    notExpired: string;
  };
  cvv: {
    required: string;
    matches: string;
  };
}
