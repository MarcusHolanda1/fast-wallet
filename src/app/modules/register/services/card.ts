import { httpClient } from '@app/shared/services/httpClient/httpClient';
import { HttpResponse } from '@app/shared/services/httpClient/types';

import { cardApi } from '../../../shared/constants/endpoints';
import { Card } from '../../../shared/types/card';

export const createCard = async (payload: {
  number: string;
  cvv: string;
  name: string;
}): Promise<HttpResponse<Card>> => {
  const response = await httpClient.post<Card>(cardApi, {
    body: payload
  });

  return response;
};
