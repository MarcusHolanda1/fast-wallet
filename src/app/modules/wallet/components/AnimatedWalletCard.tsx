import { useEffect } from 'react';
import { Dimensions, TouchableOpacity } from 'react-native';
import Animated, {
  useSharedValue,
  useAnimatedStyle,
  withTiming,
  LinearTransition
} from 'react-native-reanimated';
import CreditCard from '@app/shared/components/cards/CreditCard';
import { theme } from '@app/shared/theme/theme';
import { Card } from '@app/shared/types/card';

const STACK_OFFSET = -150;
const MOVE_DISTANCE = 40;
const DURATION_SELECTED = 260;
const DURATION_OTHER = 260;
const DURATION_RESET = 200;

interface AnimatedWalletCardProps {
  card: Card;
  index: number;
  selectedCardId: string | null;
  totalCards: number;
  colors: string[];
  onPress: () => void;
}

export default function AnimatedWalletCard({
  card,
  index,
  selectedCardId,
  totalCards,
  colors,
  onPress
}: AnimatedWalletCardProps) {
  const AnimatedTouchable = Animated.createAnimatedComponent(TouchableOpacity);
  const isSelected = selectedCardId === card.id;
  const isOther = !!selectedCardId && !isSelected;
  const backgroundColor = colors[index % colors.length];
  const offset = STACK_OFFSET;
  const screenWidth = Dimensions.get('window').width;
  const paddingVertical = 60;
  const cardWidth = screenWidth - paddingVertical;

  const translateY = useSharedValue(0);

  useEffect(() => {
    if (isOther) {
      translateY.value = -MOVE_DISTANCE / 1.5;
      translateY.value = withTiming(0, { duration: DURATION_OTHER });
    } else if (isSelected) {
      translateY.value = MOVE_DISTANCE;
      translateY.value = withTiming(0, { duration: DURATION_SELECTED });
    } else {
      translateY.value = withTiming(0, { duration: DURATION_RESET });
    }
  }, [isOther, isSelected, translateY]);

  const animatedStyle = useAnimatedStyle(() => ({
    transform: [{ translateY: translateY.value }]
  }));

  const baseStyle = isOther
    ? {
        position: 'absolute' as const,
        left: 0,
        right: 0,
        bottom: -180,
        width: cardWidth,
        height: 180,
        opacity: 0.5
      }
    : { marginTop: offset, width: cardWidth, height: 180 };

  return (
    <AnimatedTouchable
      accessibilityRole="button"
      accessibilityLabel="CartÃ£o"
      layout={LinearTransition.springify()}
      onPress={onPress}
      style={[
        baseStyle,
        { zIndex: isSelected ? 100 : totalCards + index },
        animatedStyle
      ]}
    >
      <CreditCard
        card={card}
        backgroundColor={backgroundColor}
        textColor={
          backgroundColor === theme.colors.text.black
            ? theme.colors.base.white
            : theme.colors.text.black
        }
      />
    </AnimatedTouchable>
  );
}
