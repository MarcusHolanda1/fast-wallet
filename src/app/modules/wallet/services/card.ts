import { httpClient } from '@app/shared/services/httpClient/httpClient';
import { HttpResponse } from '@app/shared/services/httpClient/types';
import { cardApi } from '@app/shared/constants/endpoints';

import { Card } from '../../../shared/types/card';

export const getCards = async (): Promise<HttpResponse<Card[]>> => {
  const response = await httpClient.get<Card[]>(cardApi);

  return response;
};
