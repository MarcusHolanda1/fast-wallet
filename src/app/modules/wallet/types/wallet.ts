import { Card } from '@app/shared/types/card';

export interface WalletState {
  selectedCard: {
    id: string | null;
    card: Card | null;
    isSelected: boolean;
  };
  nonSelectedCards: Card[];
  groupedCards: {
    selectedCard: Card | null;
    nonSelectedCards: Card[];
    hasSelection: boolean;
  };
  getCardState: (cardId: string) => {
    isSelected: boolean;
    isOther: boolean;
    shouldShow: boolean;
  };
  handleToggleSelection: (cardId: string) => void;
}
