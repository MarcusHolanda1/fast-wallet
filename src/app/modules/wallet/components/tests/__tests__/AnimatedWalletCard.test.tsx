import { fireEvent, render } from '@testing-library/react-native';
import AnimatedWalletCard from '../../AnimatedWalletCard';
import { theme } from '@app/shared/theme/theme';

import { generateFakeCardData } from '../../../../../../__mocks__/card';
import { Card } from '@app/shared/types/card';

const mockFakeCard: Card = generateFakeCardData();

const defaultProps = {
  card: mockFakeCard,
  index: 0,
  cardState: { isSelected: false, isOther: false, shouldShow: true },
  totalCards: 2,
  colors: [theme.colors.base.white, theme.colors.text.black],
  onPress: jest.fn()
};

describe('AnimatedWalletCard', () => {
  test('should render CreditCard with correct props', () => {
    const { getByLabelText } = render(<AnimatedWalletCard {...defaultProps} />);
    expect(getByLabelText('Card')).toBeTruthy();
  });

  test('should call onPress when pressed', () => {
    const onPress = jest.fn();
    const { getByTestId } = render(
      <AnimatedWalletCard {...defaultProps} onPress={onPress} />
    );
    fireEvent.press(getByTestId('animated-wallet-card'));
    expect(onPress).toHaveBeenCalled();
  });

  test('should apply correct backgroundColor and textColor', () => {
    const { getByTestId } = render(
      <AnimatedWalletCard {...defaultProps} index={1} />
    );
    expect(getByTestId('animated-wallet-card')).toBeTruthy();
  });

  test('should render with opacity 0.4 when isOther and shouldShow', () => {
    const { getByTestId } = render(
      <AnimatedWalletCard
        {...defaultProps}
        cardState={{ isSelected: false, isOther: true, shouldShow: true }}
      />
    );
    expect(getByTestId('animated-wallet-card')).toBeTruthy();
  });

  test('should render with zIndex 100 when isSelected', () => {
    const { getByTestId } = render(
      <AnimatedWalletCard
        {...defaultProps}
        cardState={{ isSelected: true, isOther: false, shouldShow: true }}
      />
    );
    expect(getByTestId('animated-wallet-card')).toBeTruthy();
  });

  test('should match snapshot', () => {
    const tree = render(<AnimatedWalletCard {...defaultProps} />);
    expect(tree.toJSON()).toMatchSnapshot();
  });
});
