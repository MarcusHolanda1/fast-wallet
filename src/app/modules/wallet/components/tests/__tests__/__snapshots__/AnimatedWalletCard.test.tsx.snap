// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AnimatedWalletCard should match snapshot 1`] = `
<View
  accessibilityLabel="Card"
  accessibilityRole="button"
  accessibilityState={
    {
      "busy": undefined,
      "checked": undefined,
      "disabled": undefined,
      "expanded": undefined,
      "selected": undefined,
    }
  }
  accessibilityValue={
    {
      "max": undefined,
      "min": undefined,
      "now": undefined,
      "text": undefined,
    }
  }
  accessible={true}
  collapsable={false}
  focusable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    {
      "height": 180,
      "marginTop": -150,
      "opacity": 1,
      "width": 690,
      "zIndex": 2,
    }
  }
  testID="animated-wallet-card"
>
  <View
    style={
      [
        {
          "borderRadius": 16,
          "elevation": 8,
          "height": 200,
          "justifyContent": "space-between",
          "paddingBottom": 40,
          "paddingHorizontal": 16,
          "paddingTop": 30,
          "shadowColor": "#000",
          "shadowOffset": {
            "height": 4,
            "width": 0,
          },
          "shadowOpacity": 0.3,
          "shadowRadius": 8,
          "width": "100%",
        },
        {
          "backgroundColor": "#FFFFFF",
        },
      ]
    }
    testID="credit-card-1"
  >
    <Text
      style={
        [
          {
            "fontSize": 18,
            "fontWeight": "400",
            "lineHeight": 20,
            "marginBottom": 3,
          },
          {
            "color": "#111111",
          },
        ]
      }
    >
      Black
    </Text>
    <View
      style={
        {
          "flex": 1,
          "justifyContent": "flex-end",
        }
      }
    >
      <Text
        style={
          [
            {
              "fontSize": 16,
              "fontWeight": "400",
              "lineHeight": 18,
              "marginBottom": 3,
            },
            {
              "color": "#111111",
            },
          ]
        }
      >
        MARCUS HOLANDA
      </Text>
      <Text
        accessibilityLabel="Final 1111"
        style={
          [
            {
              "color": "#111111",
            },
            {
              "fontSize": 14,
              "fontWeight": "400",
              "letterSpacing": 1,
              "lineHeight": 16,
              "marginBottom": 3,
            },
          ]
        }
        testID="credit-card-1-number"
      >
        •••• •••• •••• 1111
      </Text>
      <Text
        style={
          [
            {
              "fontSize": 14,
              "fontWeight": "400",
              "lineHeight": 16,
            },
            {
              "color": "#111111",
            },
          ]
        }
      >
        Validade 
        12/30
      </Text>
    </View>
  </View>
</View>
`;
